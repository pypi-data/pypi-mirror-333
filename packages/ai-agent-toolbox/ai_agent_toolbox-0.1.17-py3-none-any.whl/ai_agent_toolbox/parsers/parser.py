from typing import List
from ai_agent_toolbox.parser_event import ParserEvent

class Parser:
    def parse(self, text: str) -> List[ParserEvent]:
        pass

    def parse_chunk(self, chunk: str) -> List[ParserEvent]:
        pass

    def flush(self) -> List[ParserEvent]:
        pass
