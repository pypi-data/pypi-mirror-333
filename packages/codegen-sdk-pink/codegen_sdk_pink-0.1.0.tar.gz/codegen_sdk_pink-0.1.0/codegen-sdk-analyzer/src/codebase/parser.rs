use std::path::PathBuf;

use codegen_sdk_ast::{Definitions, References};
#[cfg(feature = "serialization")]
use codegen_sdk_common::serialize::Cache;
use codegen_sdk_cst::LANGUAGES;
use codegen_sdk_resolution::Db;
use indicatif::{ProgressBar, ProgressStyle};

use super::discovery::{FilesToParse, log_languages};
use crate::{ParsedFile, database::CodegenDatabase, parser::parse_file};
pub fn execute_op_with_progress<
    Database: Db + ?Sized + 'static,
    Input: Send + Sync,
    T: Send + Sync,
>(
    db: &Database,
    files: codegen_sdk_common::hash::FxHashSet<Input>,
    name: &str,
    parallel: bool,
    op: fn(&Database, Input) -> T,
) -> Vec<T> {
    let multi = db.multi_progress();
    let style = ProgressStyle::with_template(
        "[{elapsed_precise}] {wide_bar} {msg} [{per_sec}] [estimated time remaining: {eta}]",
    )
    .unwrap();
    let pg = multi.add(
        ProgressBar::new(files.len() as u64)
            .with_style(style)
            .with_message(name.to_string()),
    );
    let inputs = files
        .into_iter()
        .map(|file| (&pg, file, op))
        .collect::<Vec<_>>();
    let results: Vec<_> = if parallel {
        salsa::par_map(db, inputs, move |db, input| {
            let (pg, file, op) = input;
            let res = op(
                db,
                #[cfg(feature = "serialization")]
                &cache,
                file,
            );
            pg.inc(1);
            res
        })
    } else {
        inputs
            .into_iter()
            .map(|input| {
                let (pg, file, op) = input;
                let res = op(
                    db,
                    #[cfg(feature = "serialization")]
                    &cache,
                    file,
                );
                pg.inc(1);
                res
            })
            .collect()
    };
    pg.finish();
    results
}
// #[salsa::tracked]
// fn parse_files_par(db: &dyn Db, files: FilesToParse) {
//     let _: Vec<_> = execute_op_with_progress(db, files, "Parsing Files", |db, file| {
//         parse_file(db, file);
//     });
// }
#[salsa::tracked]
fn parse_files_definitions_par(db: &dyn Db, files: FilesToParse) {
    let ids = files
        .files(db)
        .iter()
        .map(|input| codegen_sdk_common::FileNodeId::new(db, input.path(db)))
        .collect::<codegen_sdk_common::hash::FxHashSet<_>>();
    let _: Vec<_> = execute_op_with_progress(db, ids, "Parsing Files", true, |db, input| {
        let file = parse_file(db, input.clone());
        if let Some(parsed) = file.file(db) {
            #[cfg(feature = "typescript")]
            if let ParsedFile::Typescript(parsed) = parsed {
                parsed.definitions(db);
                parsed.references(db);
            }
            #[cfg(feature = "python")]
            if let ParsedFile::Python(parsed) = parsed {
                parsed.definitions(db);
                parsed.references(db);
                // let deps = codegen_sdk_python::ast::dependencies(db, input);
                // for dep in deps.dependencies(db).keys() {
                //     codegen_sdk_resolution::ast::references_impl(db, dep);
                // }
            }
        }
        ()
    });
}
#[salsa::tracked]
fn compute_dependencies_par(db: &dyn Db, files: FilesToParse) {
    let ids = files
        .files(db)
        .iter()
        .map(|input| codegen_sdk_common::FileNodeId::new(db, input.path(db)))
        .collect::<codegen_sdk_common::hash::FxHashSet<_>>();
    let _targets: codegen_sdk_common::hash::FxHashSet<(PathBuf, String)> =
        execute_op_with_progress(db, ids, "Computing Dependencies", true, |db, input| {
            let file = parse_file(db, input.clone());
            if let Some(parsed) = file.file(db) {
                #[cfg(feature = "python")]
                if let ParsedFile::Python(_parsed) = parsed {
                    let deps = codegen_sdk_python::ast::dependency_keys(db, input);
                    return deps
                        .iter()
                        .map(|dep| (dep.file(db).path(db).clone(), dep.name(db).clone()))
                        .collect::<Vec<_>>();
                }
            }
            Vec::new()
        })
        .into_iter()
        .flatten()
        .collect();
    // let _: Vec<_> = execute_op_with_progress(db, targets, "Finding Usages", true, |db, input: (PathBuf, String)| {
    //     let file_node_id = codegen_sdk_common::FileNodeId::new(db, input.0);
    //     let fully_qualified_name = codegen_sdk_resolution::FullyQualifiedName::new(db, file_node_id, input.1);
    //     codegen_sdk_python::ast::references_impl(db, fully_qualified_name);
    // });
}

pub fn parse_files<'db>(
    db: &'db CodegenDatabase,
    #[cfg(feature = "serialization")] cache: &'db Cache,
    files_to_parse: FilesToParse,
) -> () {
    let _ = rayon::ThreadPoolBuilder::new()
        .stack_size(1024 * 1024 * 1024 * 10)
        .build_global();
    log_languages();
    #[cfg(feature = "serialization")]
    let cache = Cache::new().unwrap();
    #[cfg(feature = "serialization")]
    let cached = get_cached_count(&cache, &files_to_parse);
    log::info!("Parsing {} files", files_to_parse.files(db).len());
    for language in LANGUAGES.iter() {
        let mut count = 0;
        for file in files_to_parse.files(db).iter() {
            if language.should_parse(&file.path(db)).unwrap() {
                count += 1;
            }
        }
        log::info!("{} files to parse for {}", count, language.name());
    }
    parse_files_definitions_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    compute_dependencies_par(
        db,
        #[cfg(feature = "serialization")]
        &cache,
        files_to_parse,
    );
    #[cfg(feature = "serialization")]
    report_cached_count(cached, &files_to_parse.files(db));
}
