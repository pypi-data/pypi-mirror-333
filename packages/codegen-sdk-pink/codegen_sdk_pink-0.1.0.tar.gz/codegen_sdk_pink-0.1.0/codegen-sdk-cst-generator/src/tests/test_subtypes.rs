use std::collections::HashMap;

use codegen_sdk_common::parser::{Fields, Node, TypeDefinition};

use crate::{Config, generate_cst, test_util::get_language};

#[test_log::test]
fn test_basic_subtypes() {
    // Define nodes with basic subtype relationships
    let nodes = vec![
        Node {
            type_name: "expression".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "binary_expression".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "unary_expression".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "binary_expression".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "unary_expression".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];

    let language = get_language(nodes);
    let output = generate_cst(&language, Config::default()).unwrap();
    insta::assert_debug_snapshot!(crate::test_util::snapshot_string(&output));
}

#[test_log::test]
fn test_subtypes_with_fields() {
    let nodes = vec![
        Node {
            type_name: "expression".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "binary_expression".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "literal".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "binary_expression".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: Some(Fields {
                fields: HashMap::from([
                    (
                        "left".to_string(),
                        codegen_sdk_common::parser::FieldDefinition {
                            types: vec![TypeDefinition {
                                type_name: "expression".to_string(),
                                named: true,
                            }],
                            multiple: false,
                            required: true,
                        },
                    ),
                    (
                        "right".to_string(),
                        codegen_sdk_common::parser::FieldDefinition {
                            types: vec![TypeDefinition {
                                type_name: "expression".to_string(),
                                named: true,
                            }],
                            multiple: false,
                            required: true,
                        },
                    ),
                ]),
            }),
            children: None,
        },
        Node {
            type_name: "literal".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];

    let language = get_language(nodes);
    let output = generate_cst(&language, Config::default()).unwrap();
    insta::assert_debug_snapshot!(crate::test_util::snapshot_string(&output));
}

#[test_log::test]
fn test_deeply_nested_subtypes() {
    let nodes = vec![
        // Top level statement type
        Node {
            type_name: "statement".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "declaration".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "expression_statement".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Declaration with its subtypes
        Node {
            type_name: "declaration".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "function_declaration".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "class_declaration".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Function declaration with its subtype
        Node {
            type_name: "function_declaration".to_string(),
            subtypes: vec![TypeDefinition {
                type_name: "method_declaration".to_string(),
                named: true,
            }],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        // Concrete node types
        Node {
            type_name: "method_declaration".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "class_declaration".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
        Node {
            type_name: "expression_statement".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        },
    ];
    let language = get_language(nodes);
    let output = generate_cst(&language, Config::default()).unwrap();
    insta::assert_debug_snapshot!(crate::test_util::snapshot_string(&output));
}
