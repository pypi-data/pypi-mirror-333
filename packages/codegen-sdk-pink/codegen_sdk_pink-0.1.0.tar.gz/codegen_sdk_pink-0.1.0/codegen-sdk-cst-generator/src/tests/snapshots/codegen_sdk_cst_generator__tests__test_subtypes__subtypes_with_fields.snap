---
source: codegen-sdk-cst-generator/src/tests/test_subtypes.rs
expression: "crate::test_util::snapshot_string(&output)"
---
use std::sync::Arc;
use tree_sitter;
use codegen_sdk_common::*;
use subenum::subenum;
use std::backtrace::Backtrace;
use bytes::Bytes;
use ambassador::Delegate;
use derive_more::Debug;
use ambassador::delegate_to_methods;
use codegen_sdk_cst::CSTLanguage;
use crate::cst::tree::ParseContext;
use std::path::PathBuf;
#[subenum(BinaryExpressionChildren, Expression)]
#[derive(Debug, Eq, PartialEq, Clone, Hash, salsa::Update, Delegate, Ord, PartialOrd)]
#[delegate(CSTNode<'db1>)]
pub enum NodeTypes<'db1> {
    #[subenum(Expression, BinaryExpressionChildren)]
    BinaryExpression(BinaryExpression<'db1>),
    #[subenum(Expression, BinaryExpressionChildren)]
    Literal(Literal<'db1>),
}
impl<'db3> From<BinaryExpression<'db3>> for NodeTypes<'db3> {
    fn from(node: BinaryExpression<'db3>) -> Self {
        Self::BinaryExpression(node)
    }
}
impl<'db3> From<Literal<'db3>> for NodeTypes<'db3> {
    fn from(node: Literal<'db3>) -> Self {
        Self::Literal(node)
    }
}
impl<'db4> FromNode<'db4, NodeTypes<'db4>> for BinaryExpressionChildren<'db4> {
    fn from_node(
        context: &mut ParseContext<'db4, NodeTypes<'db4>>,
        node: tree_sitter::Node,
    ) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
        match node.kind_id() {
            0u16 => {
                let (node, ids) = Literal::from_node(context, node)?;
                Ok((Self::Literal(node), ids))
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
impl<'db4> FromNode<'db4, NodeTypes<'db4>> for Expression<'db4> {
    fn from_node(
        context: &mut ParseContext<'db4, NodeTypes<'db4>>,
        node: tree_sitter::Node,
    ) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
        match node.kind_id() {
            0u16 => {
                let (node, ids) = Literal::from_node(context, node)?;
                Ok((Self::Literal(node), ids))
            }
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
#[subenum(BinaryExpressionChildrenRef, ExpressionRef)]
#[derive(Debug, Eq, PartialEq, Hash, Clone, Ord, PartialOrd)]
pub enum NodeTypesRef<'db1> {
    #[subenum(ExpressionRef, BinaryExpressionChildrenRef)]
    BinaryExpression(&'db1 BinaryExpression<'db1>),
    #[subenum(ExpressionRef, BinaryExpressionChildrenRef)]
    Literal(&'db1 Literal<'db1>),
}
impl<'db3> From<&'db3 BinaryExpression<'db3>> for NodeTypesRef<'db3> {
    fn from(node: &'db3 BinaryExpression<'db3>) -> Self {
        Self::BinaryExpression(node)
    }
}
impl<'db3> From<&'db3 Literal<'db3>> for NodeTypesRef<'db3> {
    fn from(node: &'db3 Literal<'db3>) -> Self {
        Self::Literal(node)
    }
}
impl<'db3> NodeTypes<'db3> {
    pub fn as_ref(&'db3 self) -> NodeTypesRef<'db3> {
        match self {
            Self::BinaryExpression(data) => NodeTypesRef::BinaryExpression(data),
            Self::Literal(data) => NodeTypesRef::Literal(data),
        }
    }
}
#[delegate_to_methods]
#[delegate(CSTNode<'db3>, target_ref = "deref")]
impl<'db3> NodeTypesRef<'db3> {
    fn deref<'db2>(&'db2 self) -> &'db2 dyn CSTNode<'db3> {
        match self {
            Self::BinaryExpression(data) => *data,
            Self::Literal(data) => *data,
        }
    }
}
impl<'db3> From<&'db3 NodeTypes<'db3>> for NodeTypesRef<'db3> {
    fn from(node: &'db3 NodeTypes<'db3>) -> Self {
        node.as_ref().into()
    }
}
impl<'db3> From<NodeTypesRef<'db3>> for NodeTypes<'db3> {
    fn from(node: NodeTypesRef<'db3>) -> Self {
        match node {
            NodeTypesRef::BinaryExpression(data) => {
                Self::BinaryExpression((*data).clone())
            }
            NodeTypesRef::Literal(data) => Self::Literal((*data).clone()),
        }
    }
}
impl<'db3> From<&'db3 NodeTypesRef<'db3>> for NodeTypes<'db3> {
    fn from(node: &'db3 NodeTypesRef<'db3>) -> Self {
        match node {
            NodeTypesRef::BinaryExpression(data) => {
                Self::BinaryExpression((*data).clone())
            }
            NodeTypesRef::Literal(data) => Self::Literal((*data).clone()),
        }
    }
}
impl<'db3> TryFrom<NodeTypesRef<'db3>> for &'db3 BinaryExpression<'db3> {
    type Error = codegen_sdk_cst::ConversionError;
    fn try_from(node: NodeTypesRef<'db3>) -> Result<Self, Self::Error> {
        if let NodeTypesRef::BinaryExpression(node) = node {
            Ok(node)
        } else {
            Err(codegen_sdk_cst::ConversionError {
                expected: "TODO".to_string(),
                actual: node.kind_name().to_string(),
                backtrace: Backtrace::capture(),
            })
        }
    }
}
impl<'db3> TryFrom<NodeTypesRef<'db3>> for &'db3 Literal<'db3> {
    type Error = codegen_sdk_cst::ConversionError;
    fn try_from(node: NodeTypesRef<'db3>) -> Result<Self, Self::Error> {
        if let NodeTypesRef::Literal(node) = node {
            Ok(node)
        } else {
            Err(codegen_sdk_cst::ConversionError {
                expected: "TODO".to_string(),
                actual: node.kind_name().to_string(),
                backtrace: Backtrace::capture(),
            })
        }
    }
}
impl<'db3> BinaryExpressionChildren<'db3> {
    pub fn as_ref(&'db3 self) -> BinaryExpressionChildrenRef<'db3> {
        match self {
            Self::BinaryExpression(data) => {
                BinaryExpressionChildrenRef::BinaryExpression(data)
            }
            Self::Literal(data) => BinaryExpressionChildrenRef::Literal(data),
        }
    }
}
#[delegate_to_methods]
#[delegate(CSTNode<'db3>, target_ref = "deref")]
impl<'db3> BinaryExpressionChildrenRef<'db3> {
    fn deref<'db2>(&'db2 self) -> &'db2 dyn CSTNode<'db3> {
        match self {
            Self::BinaryExpression(data) => *data,
            Self::Literal(data) => *data,
        }
    }
}
impl<'db3> From<&'db3 BinaryExpressionChildren<'db3>> for NodeTypesRef<'db3> {
    fn from(node: &'db3 BinaryExpressionChildren<'db3>) -> Self {
        node.as_ref().into()
    }
}
impl<'db3> From<BinaryExpressionChildrenRef<'db3>> for BinaryExpressionChildren<'db3> {
    fn from(node: BinaryExpressionChildrenRef<'db3>) -> Self {
        match node {
            BinaryExpressionChildrenRef::BinaryExpression(data) => {
                Self::BinaryExpression((*data).clone())
            }
            BinaryExpressionChildrenRef::Literal(data) => Self::Literal((*data).clone()),
        }
    }
}
impl<'db3> From<&'db3 BinaryExpressionChildrenRef<'db3>>
for BinaryExpressionChildren<'db3> {
    fn from(node: &'db3 BinaryExpressionChildrenRef<'db3>) -> Self {
        match node {
            BinaryExpressionChildrenRef::BinaryExpression(data) => {
                Self::BinaryExpression((*data).clone())
            }
            BinaryExpressionChildrenRef::Literal(data) => Self::Literal((*data).clone()),
        }
    }
}
impl<'db3> TryFrom<BinaryExpressionChildrenRef<'db3>> for &'db3 BinaryExpression<'db3> {
    type Error = codegen_sdk_cst::ConversionError;
    fn try_from(node: BinaryExpressionChildrenRef<'db3>) -> Result<Self, Self::Error> {
        if let BinaryExpressionChildrenRef::BinaryExpression(node) = node {
            Ok(node)
        } else {
            Err(codegen_sdk_cst::ConversionError {
                expected: "TODO".to_string(),
                actual: node.kind_name().to_string(),
                backtrace: Backtrace::capture(),
            })
        }
    }
}
impl<'db3> TryFrom<BinaryExpressionChildrenRef<'db3>> for &'db3 Literal<'db3> {
    type Error = codegen_sdk_cst::ConversionError;
    fn try_from(node: BinaryExpressionChildrenRef<'db3>) -> Result<Self, Self::Error> {
        if let BinaryExpressionChildrenRef::Literal(node) = node {
            Ok(node)
        } else {
            Err(codegen_sdk_cst::ConversionError {
                expected: "TODO".to_string(),
                actual: node.kind_name().to_string(),
                backtrace: Backtrace::capture(),
            })
        }
    }
}
impl<'db3> Expression<'db3> {
    pub fn as_ref(&'db3 self) -> ExpressionRef<'db3> {
        match self {
            Self::BinaryExpression(data) => ExpressionRef::BinaryExpression(data),
            Self::Literal(data) => ExpressionRef::Literal(data),
        }
    }
}
#[delegate_to_methods]
#[delegate(CSTNode<'db3>, target_ref = "deref")]
impl<'db3> ExpressionRef<'db3> {
    fn deref<'db2>(&'db2 self) -> &'db2 dyn CSTNode<'db3> {
        match self {
            Self::BinaryExpression(data) => *data,
            Self::Literal(data) => *data,
        }
    }
}
impl<'db3> From<&'db3 Expression<'db3>> for NodeTypesRef<'db3> {
    fn from(node: &'db3 Expression<'db3>) -> Self {
        node.as_ref().into()
    }
}
impl<'db3> From<ExpressionRef<'db3>> for Expression<'db3> {
    fn from(node: ExpressionRef<'db3>) -> Self {
        match node {
            ExpressionRef::BinaryExpression(data) => {
                Self::BinaryExpression((*data).clone())
            }
            ExpressionRef::Literal(data) => Self::Literal((*data).clone()),
        }
    }
}
impl<'db3> From<&'db3 ExpressionRef<'db3>> for Expression<'db3> {
    fn from(node: &'db3 ExpressionRef<'db3>) -> Self {
        match node {
            ExpressionRef::BinaryExpression(data) => {
                Self::BinaryExpression((*data).clone())
            }
            ExpressionRef::Literal(data) => Self::Literal((*data).clone()),
        }
    }
}
impl<'db3> TryFrom<ExpressionRef<'db3>> for &'db3 BinaryExpression<'db3> {
    type Error = codegen_sdk_cst::ConversionError;
    fn try_from(node: ExpressionRef<'db3>) -> Result<Self, Self::Error> {
        if let ExpressionRef::BinaryExpression(node) = node {
            Ok(node)
        } else {
            Err(codegen_sdk_cst::ConversionError {
                expected: "TODO".to_string(),
                actual: node.kind_name().to_string(),
                backtrace: Backtrace::capture(),
            })
        }
    }
}
impl<'db3> TryFrom<ExpressionRef<'db3>> for &'db3 Literal<'db3> {
    type Error = codegen_sdk_cst::ConversionError;
    fn try_from(node: ExpressionRef<'db3>) -> Result<Self, Self::Error> {
        if let ExpressionRef::Literal(node) = node {
            Ok(node)
        } else {
            Err(codegen_sdk_cst::ConversionError {
                expected: "TODO".to_string(),
                actual: node.kind_name().to_string(),
                backtrace: Backtrace::capture(),
            })
        }
    }
}
#[derive(Debug, Eq, PartialEq, Clone, salsa::Update)]
pub struct BinaryExpression<'db> {
    id: CSTNodeId<'db>,
    file_id: FileNodeId,
    start_byte: usize,
    end_byte: usize,
    start_position: Point<'db>,
    end_position: Point<'db>,
    #[debug(ignore)]
    buffer: Arc<Bytes>,
    is_error: bool,
    pub left: indextree::NodeId,
    pub right: indextree::NodeId,
}
impl<'db> FromNode<'db, NodeTypes<'db>> for BinaryExpression<'db> {
    fn from_node(
        context: &mut ParseContext<'db, NodeTypes<'db>>,
        node: tree_sitter::Node,
    ) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
        let start_position = Point::from(context.db, node.start_position());
        let end_position = Point::from(context.db, node.end_position());
        let id = CSTNodeId::new(context.db, context.file_id, node.id(), context.root);
        let mut ids = Vec::new();
        let left = get_child_by_field_name::<
            NodeTypes<'db>,
            Expression<'db>,
        >(context, &node, "left")?;
        ids.push(left.clone());
        let right = get_child_by_field_name::<
            NodeTypes<'db>,
            Expression<'db>,
        >(context, &node, "right")?;
        ids.push(right.clone());
        Ok((
            Self {
                start_byte: node.start_byte(),
                end_byte: node.end_byte(),
                start_position: start_position,
                end_position: end_position,
                buffer: context.buffer.clone(),
                is_error: node.is_error(),
                id,
                file_id: context.file_id.clone(),
                left,
                right,
            },
            ids,
        ))
    }
}
impl<'db> BinaryExpression<'db> {
    const KIND_NAME: &'static str = "binary_expression";
    pub fn left(&self, tree: &'db Tree<NodeTypes<'db>>) -> ExpressionRef<'db> {
        tree.get(&self.left).unwrap().as_ref().try_into().unwrap()
    }
    pub fn right(&self, tree: &'db Tree<NodeTypes<'db>>) -> ExpressionRef<'db> {
        tree.get(&self.right).unwrap().as_ref().try_into().unwrap()
    }
}
impl<'db> CSTNode<'db> for BinaryExpression<'db> {
    fn kind_name(&self) -> &str {
        &Self::KIND_NAME
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        0u16
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        true
    }
    fn id(&self) -> CSTNodeId<'db> {
        self.id
    }
    fn file_id(&self) -> FileNodeId {
        self.file_id
    }
}
impl<'db1> HasChildren<'db1, NodeTypes<'db1>> for BinaryExpression<'db1> {
    type Child<'db2> = BinaryExpressionChildrenRef<'db2> where Self: 'db2;
    fn children<'db2>(
        &'db2 self,
        context: &'db2 Tree<NodeTypes<'db2>>,
    ) -> Vec<Self::Child<'db2>> {
        let mut children: Vec<Self::Child<'db2>> = vec![];
        children.push(context.get(&self.left).unwrap().as_ref().try_into().unwrap());
        children.push(context.get(&self.right).unwrap().as_ref().try_into().unwrap());
        children.sort_by_key(|c| c.start_byte());
        children
    }
    fn children_by_field_name<'db2>(
        &'db2 self,
        context: &'db2 Tree<NodeTypes<'db2>>,
        field_name: &str,
    ) -> Vec<Self::Child<'db2>> {
        match field_name {
            "left" => {
                vec![context.get(& self.left).unwrap().as_ref().try_into().unwrap()]
            }
            "right" => {
                vec![context.get(& self.right).unwrap().as_ref().try_into().unwrap()]
            }
            _ => vec![],
        }
    }
    fn children_by_field_id<'db2>(
        &'db2 self,
        context: &'db2 Tree<NodeTypes<'db2>>,
        field_id: u16,
    ) -> Vec<Self::Child<'db2>> {
        match field_id {
            1u16 => vec![context.get(& self.left).unwrap().as_ref().try_into().unwrap()],
            1u16 => vec![context.get(& self.right).unwrap().as_ref().try_into().unwrap()],
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for BinaryExpression<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
impl<'db> PartialOrd for BinaryExpression<'db> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl<'db> Ord for BinaryExpression<'db> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        let res = self.start_byte().cmp(&other.start_byte());
        if res == std::cmp::Ordering::Equal {
            self.end_byte().cmp(&other.end_byte())
        } else {
            res
        }
    }
}
#[derive(Debug, Eq, PartialEq, Clone, salsa::Update)]
pub struct Literal<'db> {
    id: CSTNodeId<'db>,
    file_id: FileNodeId,
    start_byte: usize,
    end_byte: usize,
    start_position: Point<'db>,
    end_position: Point<'db>,
    #[debug(ignore)]
    buffer: Arc<Bytes>,
    is_error: bool,
}
impl<'db> FromNode<'db, NodeTypes<'db>> for Literal<'db> {
    fn from_node(
        context: &mut ParseContext<'db, NodeTypes<'db>>,
        node: tree_sitter::Node,
    ) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
        let start_position = Point::from(context.db, node.start_position());
        let end_position = Point::from(context.db, node.end_position());
        let id = CSTNodeId::new(context.db, context.file_id, node.id(), context.root);
        let mut ids = Vec::new();
        Ok((
            Self {
                start_byte: node.start_byte(),
                end_byte: node.end_byte(),
                start_position: start_position,
                end_position: end_position,
                buffer: context.buffer.clone(),
                is_error: node.is_error(),
                id,
                file_id: context.file_id.clone(),
            },
            ids,
        ))
    }
}
impl<'db> Literal<'db> {
    const KIND_NAME: &'static str = "literal";
}
impl<'db> CSTNode<'db> for Literal<'db> {
    fn kind_name(&self) -> &str {
        &Self::KIND_NAME
    }
    fn start_byte(&self) -> usize {
        self.start_byte
    }
    fn end_byte(&self) -> usize {
        self.end_byte
    }
    fn start_position(&self) -> Point<'db> {
        self.start_position
    }
    fn end_position(&self) -> Point<'db> {
        self.end_position
    }
    fn buffer(&self) -> &Bytes {
        &self.buffer
    }
    fn kind_id(&self) -> u16 {
        0u16
    }
    fn is_error(&self) -> bool {
        self.is_error
    }
    fn is_named(&self) -> bool {
        true
    }
    fn id(&self) -> CSTNodeId<'db> {
        self.id
    }
    fn file_id(&self) -> FileNodeId {
        self.file_id
    }
}
impl<'db1> HasChildren<'db1, NodeTypes<'db1>> for Literal<'db1> {
    type Child<'db2> = Literal<'db2> where Self: 'db2;
    fn children<'db2>(
        &'db2 self,
        context: &'db2 Tree<NodeTypes<'db2>>,
    ) -> Vec<Self::Child<'db2>> {
        vec![]
    }
    fn children_by_field_name<'db2>(
        &'db2 self,
        context: &'db2 Tree<NodeTypes<'db2>>,
        field_name: &str,
    ) -> Vec<Self::Child<'db2>> {
        match field_name {
            _ => vec![],
        }
    }
    fn children_by_field_id<'db2>(
        &'db2 self,
        context: &'db2 Tree<NodeTypes<'db2>>,
        field_id: u16,
    ) -> Vec<Self::Child<'db2>> {
        match field_id {
            _ => vec![],
        }
    }
}
impl<'db> std::hash::Hash for Literal<'db> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}
impl<'db> PartialOrd for Literal<'db> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl<'db> Ord for Literal<'db> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        let res = self.start_byte().cmp(&other.start_byte());
        if res == std::cmp::Ordering::Equal {
            self.end_byte().cmp(&other.end_byte())
        } else {
            res
        }
    }
}
impl<'db> TreeNode for NodeTypes<'db> {}
#[salsa::tracked]
pub struct Parsed<'db> {
    #[id]
    id: FileNodeId,
    #[tracked]
    #[return_ref]
    #[no_clone]
    #[no_eq]
    pub tree: Arc<Tree<NodeTypes<'db>>>,
    pub program: indextree::NodeId,
}
pub fn parse_program_raw<'db>(
    db: &'db dyn salsa::Database,
    input: codegen_sdk_cst::File,
) -> Option<Parsed<'db>> {
    let buffer = Bytes::from(input.content(db).as_bytes().to_vec());
    let tree = codegen_sdk_common::language::language::Language
        .parse_tree_sitter(&input.content(db));
    match tree {
        Ok(tree) => {
            if tree.root_node().has_error() {
                ParseError::SyntaxError.report(db);
                None
            } else {
                let mut context = ParseContext::new(
                    db,
                    input.path(db),
                    input.root(db),
                    buffer,
                );
                let root_id = Program::orphaned(&mut context, tree.root_node())
                    .map_or_else(
                        |e| {
                            e.report(db);
                            None
                        },
                        |program| { Some(program) },
                    );
                if let Some(program) = root_id {
                    Some(
                        Parsed::new(db, context.file_id, Arc::new(context.tree), program),
                    )
                } else {
                    None
                }
            }
        }
        Err(e) => {
            e.report(db);
            None
        }
    }
}
#[salsa::tracked(return_ref)]
pub fn parse_program(
    db: &dyn salsa::Database,
    input: codegen_sdk_cst::File,
) -> Parsed<'_> {
    let raw = parse_program_raw(db, input);
    if let Some(parsed) = raw {
        parsed
    } else {
        panic!("Failed to parse program");
    }
}
pub struct Language;
impl CSTLanguage for Language {
    type Types<'db> = NodeTypes<'db>;
    type Program<'db> = Program<'db>;
    fn language() -> &'static codegen_sdk_common::language::Language {
        &codegen_sdk_common::language::language::Language
    }
    fn parse<'db>(
        db: &'db dyn salsa::Database,
        content: std::string::String,
    ) -> Option<
        (&'db Self::Program<'db>, &'db Tree<Self::Types<'db>>, indextree::NodeId),
    > {
        let input = codegen_sdk_cst::File::new(
            db,
            std::path::PathBuf::new(),
            content,
            std::path::PathBuf::new(),
        );
        let parsed = parse_program(db, input);
        let program_id = parsed.program(db);
        let tree = parsed.tree(db);
        let program = tree.get(&program_id).unwrap().as_ref();
        Some((program.try_into().unwrap(), tree, program_id))
    }
}
